### ISSUE TITLE
+c interfaces returning +o +j

### ISSUE THREAD

**Issue by mknejp** [dropbox/djinni#26]
Hi,

the current code generation is broken for cases where a +c interface returns +j or +o interfaces.

Given:
````
extern = interface +j +o {...}
intern = interface +c { foo(): extern; }
````
On the JNI side this tries to call `Extern::toJava()`, which does not exist, and for Obj-C it tries to `static_pointer_cast` from `Extern` to `ExternObjcProxy`, which is mostly likely not the case and causes undefined behavior.

I think the easiest fix would be to disallow returning +j +o interfaces from +c. Supporting this properly would be a bit more involved I think.

**Comment by j4cbo** [comment]
This is fixed now, right?

**Comment by mknejp** [comment]
Well, there is an `assert(dynamic_cast...)` in there now, so if someone screws up it shows up in a debug build, but with asserts disabled it's just a plain `static_cast`. Question is does this suffice or should `dynamic_cast` stay there and make a fallback solution kick in if it fails.

**Comment by j4cbo** [comment]
I think what we should probably do is `dynamic_cast` a reference, so we'll get a `bad_cast` if the type is wrong (essentially, always asserting). I made that change in my proposed fix for #81 - in particular, see the new wrapping/unwrapping code here: https://github.com/dropbox/djinni/commit/f0aec36d99764ca62903ba58fbb9c4af9c886c54#diff-61477e190b9e4951838edf1fa955dd19R216

This mirrors the existing Java scheme, though with slightly less template trickery (see https://github.com/dropbox/djinni/blob/master/support-lib/jni/djinni_support.hpp#L438 and particularly _unwrapJavaProxy)

Closing this out; let's discuss in #81.
